### ISSUE TITLE
Validation loss is not calculated correctly when using multiple devices

### ISSUE THREAD

**Issue by konstantinjdobler** [Lightning-AI/lit-gpt#599]
We're calling `validate` on all devices that are used for training, therefore each process only sees a fraction of the whole validation set. In all the finetune and pretrain scripts, we never reduce the resulting `val_loss` across devices => what gets logged / printed is actually only the `val_loss` on the data shard of the first process. If you're unlucky / your dataset is not homogenous, this could potentially be quite misleading. Also, it leads to different results depending on the number of devices used.

The good news is it should be fixed with a simple all_reduce across processes:
```python
val_loss = fabric.all_reduce(val_loss, reduce_op="mean")
```
 
For a single process this should be a no-op (?) but if not we can also if-guard depending on the world size. I'm happy to open a PR if you agree.

**Comment by awaelchli** [comment]
The validation loss is computed as an approximate over a few samples (eval_iters). The dataset is not sharded, and the random sampling is done with replacement essentially (see get_batch). Adding the reduction won't make it more correct. For simplicity, we just decided to follow the nano-gpt approach of approximating the val loss, that's it.

If you're adopting these scripts, you have a few options: 1) you can increase val_iters 2) you can do the reduction 3) you can iterate over the dataset without random sampling 4) you can wrap the logic in a dataloader and use `fabric.setup_dataloaders()` to get sharded data loading and iterate over that (then reduce across devices).

**Comment by konstantinjdobler** [comment]
I am mostly adapting the pretrain scripts (`red pajama.py & `openwebtext.py`), where I do believe the traditional `DataLoader`s w/ sharded datasets across processes are used. I see that it's a different case for the fine-tune scripts if all the sampling is random anyways.

I did 4) already in my adapted version but I understand if you want to keep these scripts as lightweight as possible.
